mlflow/telemetry/client.py

```diff
@@ -1,33 +1,214 @@
+        self._atexit_callback_registered = False
+
+        # Track consumer threads
+        self._consumer_threads = []
+
+        self._batch_size = BATCH_SIZE
+        self._batch_time_interval = BATCH_TIME_INTERVAL
+        self._pending_records: list[TelemetryRecord] = []
+        self._last_batch_time = time.time()
+        self._batch_lock = threading.Lock()
+
+    def add_record(self, record: APIRecord):
+        """
+        Add a record to be batched and sent to the telemetry server.
+        """
+        if not self.is_active:
+            self.activate()
+
+        if self._is_stopped:
+            _logger.debug("Telemetry is stopped, skipping adding record")
+            return
+
+        data = self._generate_telemetry_record(record)
+        with self._batch_lock:
+            self._pending_records.append(data)
+
+            should_send = (
+                len(self._pending_records) >= self._batch_size
+                or (time.time() - self._last_batch_time) >= self._batch_time_interval
+            )
+
+            if should_send:
+                self._send_batch()
+
+    def _send_batch(self):
+        """Send the current batch of records."""
+        if not self._pending_records:
+            return
+
+        # Create a copy of the current batch and clear the pending list
+        self._last_batch_time = time.time()
+
+        try:
+            self._queue.put(self._pending_records, block=False)
+            self._pending_records = []
+        except Full:
+            # TODO: record this case
+            _logger.debug("Telemetry queue is full, skipping sending data.")
+
+    def _process_records(self, records: list[TelemetryRecord]):
+        """Process a batch of telemetry records."""
+        try:
+            data = {
+                "records": [
+                    {"data": record.data, "partition-key": record.partition_key}
+                    for record in records
+                ]
+            }
+            response = requests.post(
+                self.telemetry_url,
+                json=data,
+                headers={"Content-Type": "application/json"},
+            )
+            if response.status_code != 200:
+                _logger.debug(
+                    f"Failed to send telemetry records. Status code: {response.status_code}, "
+                    f"Response: {response.text}"
+                )
+        except Exception as e:
+            _logger.debug(
+                f"Failed to process telemetry records. Error: {e}.",
+                exc_info=True,
+            )
 
-    # temp helper function to test
-    # TODO: update this in next PR
-    def add_record(self, record: Record):
+    def _consumer(self) -> None:
+        """Individual consumer that processes records from the queue."""
+        while not self._is_stopped:
+            try:
+                records = self._queue.get(timeout=1)
+            except Empty:
+                continue
+
+            self._process_records(records)
```

Comment:
This looks a bit scary. `records` is undefined if `_is_stopped` is falsy.

Reply:
If _is_stopped is True, this won't be executed; if it's False, records is always fetched first, and this is skipped if no records. Could you elaborate on this?

Reply:
My bad. I thought `self._process_records(records)` was unindented